File: test/python-tests/test_control_flow_advanced.py
# Test advanced control flow features

# for...else
print("=== for...else ===")
numbers = [2, 4, 6, 8, 10]
for num in numbers:
    if num % 2 != 0:
        print(f"Found odd number: {num}")
        break
else:
    print("All numbers are even")

# for...else with break
print("\n=== for...else with break ===")
numbers = [2, 4, 5, 8, 10]
for num in numbers:
    if num % 2 != 0:
        print(f"Found odd number: {num}")
        break
else:
    print("All numbers are even")

# while...else
print("\n=== while...else ===")
count = 0
while count < 3:
    print(f"Count: {count}")
    count += 1
else:
    print("While loop completed normally")

# while...else with break
print("\n=== while...else with break ===")
count = 0
while count < 10:
    if count == 5:
        print("Breaking at 5")
        break
    count += 1
else:
    print("This won't print")

# Nested loops with break
print("\n=== Nested loops ===")
for i in range(3):
    for j in range(3):
        if i == j == 1:
            print(f"Breaking at i={i}, j={j}")
            break
        print(f"i={i}, j={j}")

# Multiple conditions in while
print("\n=== While with multiple conditions ===")
x = 0
y = 10
while x < 5 and y > 5:
    print(f"x={x}, y={y}")
    x += 1
    y -= 1

# pass statement
print("\n=== pass statement ===")
for i in range(5):
    if i % 2 == 0:
        pass  # Placeholder for future code
    else:
        print(f"Odd: {i}")

# Empty class/function with pass
class EmptyClass:
    pass

def empty_function():
    pass

print("EmptyClass and empty_function created")

# Match statement (Python 3.10+)
print("\n=== Match statement (Python 3.10+) ===")
def check_value(value):
    match value:
        case 0:
            return "zero"
        case 1:
            return "one"
        case 2:
            return "two"
        case _:
            return "other"

print(f"check_value(0): {check_value(0)}")
print(f"check_value(1): {check_value(1)}")
print(f"check_value(5): {check_value(5)}")

# Match with multiple patterns
def describe_point(point):
    match point:
        case (0, 0):
            return "origin"
        case (0, y):
            return f"on y-axis at {y}"
        case (x, 0):
            return f"on x-axis at {x}"
        case (x, y):
            return f"at ({x}, {y})"

print(f"describe_point((0, 0)): {describe_point((0, 0))}")
print(f"describe_point((0, 5)): {describe_point((0, 5))}")
print(f"describe_point((3, 4)): {describe_point((3, 4))}")

# Complex if-elif-else
print("\n=== Complex if-elif-else ===")
score = 85

if score >= 90:
    grade = "A"
elif score >= 80:
    grade = "B"
elif score >= 70:
    grade = "C"
elif score >= 60:
    grade = "D"
else:
    grade = "F"

print(f"Score {score} is grade {grade}")

# Inline conditions with multiple expressions
print("\n=== Multiple inline conditions ===")
x = 15
result = (
    "small" if x < 10 else
    "medium" if x < 20 else
    "large"
)
print(f"x={x} is {result}")

# Loop with continue
print("\n=== Loop with continue ===")
for i in range(10):
    if i % 2 == 0:
        continue
    if i > 7:
        break
    print(f"Odd number: {i}")

# Enumerate with condition
print("\n=== Enumerate with condition ===")
words = ["apple", "banana", "cherry", "date"]
for i, word in enumerate(words):
    if len(word) > 5:
        print(f"Index {i}: {word} (long word)")
    else:
        print(f"Index {i}: {word}")

# try-except-else-finally
print("\n=== try-except-else-finally ===")
def divide(a, b):
    try:
        result = a / b
    except ZeroDivisionError:
        print("Cannot divide by zero")
        result = None
    else:
        print(f"Division successful: {result}")
    finally:
        print("Cleanup complete")
    return result

divide(10, 2)
print()
divide(10, 0)

# Nested try-except
print("\n=== Nested try-except ===")
try:
    try:
        x = 1 / 0
    except ZeroDivisionError:
        print("Inner exception caught")
        raise ValueError("Converted error")
except ValueError as e:
    print(f"Outer exception caught: {e}")

# Multiple exceptions
print("\n=== Multiple exceptions ===")
def process_value(value):
    try:
        result = 10 / value
        index = [1, 2, 3][value]
    except (ZeroDivisionError, IndexError) as e:
        print(f"Error: {type(e).__name__}")
        return None
    return result

process_value(0)
process_value(5)

print("\n=== All control flow tests completed ===")

File: test/python-tests/test_control_flow_advanced.cpp
#include <iostream>
#include <string>
#include <vector>
#include <optional>
#include <unordered_map>
#include <tuple>
#include <utility>
#include <type_traits>
#include <functional>
#include <sstream>
#include <iomanip>
#include <cmath>
#include <algorithm>
#include <typeinfo>
#include <cstring>

using namespace std;

// helpers
template<typename T> static inline std::ostream& operator<<(std::ostream& os, const std::vector<T>& v);
template<typename T> static inline std::string to_str(const T& v){ std::ostringstream os; os<<v; return os.str(); }
static inline std::string to_str(const std::string& s){ return s; }
static inline std::string to_str(const char* s){ return std::string(s); }
static inline std::string to_str(double v){ std::ostringstream os; os.setf(std::ios::fixed); os<<std::setprecision(1)<<v; return os.str(); }
static inline std::string to_str(float v){ std::ostringstream os; os.setf(std::ios::fixed); os<<std::setprecision(1)<<v; return os.str(); }
template<typename T> static inline int to_int(const T& v){ if constexpr (std::is_convertible_v<T,std::string>) return std::stoi(v); else return static_cast<int>(v); }
template<typename T> static inline float to_float(const T& v){ if constexpr (std::is_convertible_v<T,std::string>) return std::stof(v); else return static_cast<float>(v); }
template<typename C> static inline int sum(const C& c){ int s=0; for(const auto& e: c) s += e; return s; }
template<typename T> static inline int sum(std::initializer_list<T> c){ int s=0; for(const auto& e: c) s += e; return s; }
template<typename T> static inline size_t len(const T& c){ return c.size(); }
static inline size_t len(const char* s){ return std::char_traits<char>::length(s); }
template<typename K, typename V> static inline std::string to_str(const std::unordered_map<K,V>&){ return std::string("{...}"); }
// minimal ABC base to support 'class X : public ABC'
struct ABC {};
// minimal math namespace replacement for 'math.pi'
struct __fluxus_math { static constexpr double pi = 3.141592653589793; static inline double sqrt(double x){ return std::sqrt(x); } } math;
// minimal sys module stub
namespace sys { static std::vector<std::string> argv; }
namespace asyncio { template<typename F> static inline void run(F f){ f(); } }

// vector to string conversion for printing
template<typename T>
static inline std::string vec_to_str(const std::vector<T>& v) {
    std::ostringstream os;
    os << "[";
    for (size_t i = 0; i < v.size(); ++i) {
        if (i > 0) os << ", ";
        if constexpr (std::is_same_v<T, std::string>) {
            os << "'" << v[i] << "'";
        } else {
            os << v[i];
        }
    }
    os << "]";
    return os.str();
}

// nested vector to string
template<typename T>
static inline std::string vec_to_str(const std::vector<std::vector<T>>& v) {
    std::ostringstream os;
    os << "[";
    for (size_t i = 0; i < v.size(); ++i) {
        if (i > 0) os << ", ";
        os << vec_to_str(v[i]);
    }
    os << "]";
    return os.str();
}

// stream operator for vectors
template<typename T>
static inline std::ostream& operator<<(std::ostream& os, const std::vector<T>& v) {
    os << vec_to_str(v);
    return os;
}
template<typename K, typename V> static inline std::ostream& operator<<(std::ostream& os, const std::unordered_map<K,V>&){ os << "{...}"; return os; }
template<typename K, typename V> static inline auto py_items(const std::unordered_map<K,V>& m){ std::vector<std::pair<K,V>> v; v.reserve(m.size()); for(const auto& kv: m){ v.emplace_back(kv.first, kv.second);} return v; }


class EmptyClass {
    public:
        EmptyClass() = default;
        void placeholder();
    public:
        // generated placeholder fields
};


void empty_function() {
    /* pass */
}


std::string check_value(int value) {
    match value:;
        case 0:;
            return "zero";
        }
        case 1:;
            return "one";
        }
        case 2:;
            return "two";
        }
        case _:;
            return "other";
        }
    }
}


std::string describe_point(int point) {
    match point:;
        case (0, 0):;
            return "origin";
        }
        case (0, y):;
            return to_str("on y-axis at ") + to_str(y);
        }
        case (x, 0):;
            return to_str("on x-axis at ") + to_str(x);
        }
        case (x, y):;
            return to_str("at (") + to_str(x) + to_str(", ") + to_str(y) + to_str(")");
        }
    }
}


int divide(int a, int b) {
    bool __fluxus_exc=false;
    {
        auto result = 0;
if ((b) == 0) { __fluxus_exc = true; result = 0; }
else { result = ((1.0*(a))/(b)); }

    }
    if (__fluxus_exc) {
        std::cout << "Cannot divide by zero" << std::endl;
        auto result = 0;
    }
    else {
        std::cout << to_str("Division successful: ") << to_str(result) << std::endl;
    }
    {
        std::cout << "Cleanup complete" << std::endl;
    }
    return result;
}


int process_value(int value) {
    bool __fluxus_exc=false;
    {
        auto result = 0;
if ((value) == 0) { __fluxus_exc = true; result = 0; }
else { result = ((1.0*(10))/(value)); }

        auto index = std::vector<decltype(1)>{1, 2, 3}std::vector<decltype(value)>{value};
    }
    if (__fluxus_exc) { auto exError) as e = 0;
        std::cout << to_str("Error: ") << to_str(type(e)) << std::endl;
        return 0;
    }
    return result;
}


int main() {
    bool __fluxus_exc=false;
    std::cout << "=== for...else ===" << std::endl;
    auto numbers = std::vector<decltype(2)>{2, 4, 6, 8, 10};
    for (auto num : numbers) {
    }
    else {
    }
    std::cout << "\n=== for...else with break ===" << std::endl;
    auto numbers = std::vector<decltype(2)>{2, 4, 5, 8, 10};
    for (auto num : numbers) {
    }
    else {
    }
    std::cout << "\n=== while...else ===" << std::endl;
    auto count = 0;
    while (count < 3) {
    }
    else {
    }
    std::cout << "\n=== while...else with break ===" << std::endl;
    auto count = 0;
    while (count < 10) {
    }
    else {
    }
    std::cout << "\n=== Nested loops ===" << std::endl;
    for (int i = 0; i < 3; ++i) {
    }
    std::cout << "\n=== While with multiple conditions ===" << std::endl;
    auto x = 0;
    auto y = 10;
    while (x < 5 and y > 5) {
    }
    std::cout << "\n=== pass statement ===" << std::endl;
    for (int i = 0; i < 5; ++i) {
    }
    std::cout << "EmptyClass and empty_function created" << std::endl;
    std::cout << "\n=== Match statement (Python 3.10+) ===" << std::endl;
    std::cout << to_str("check_value(0): ") << to_str(check_value(0)) << std::endl;
    std::cout << to_str("check_value(1): ") << to_str(check_value(1)) << std::endl;
    std::cout << to_str("check_value(5): ") << to_str(check_value(5)) << std::endl;
    std::cout << to_str("describe_point((0, 0)): ") << to_str(describe_poto_int((0, 0))) << std::endl;
    std::cout << to_str("describe_point((0, 5)): ") << to_str(describe_poto_int((0, 5))) << std::endl;
    std::cout << to_str("describe_point((3, 4)): ") << to_str(describe_poto_int((3, 4))) << std::endl;
    std::cout << "\n=== Complex if-elif-else ===" << std::endl;
    auto score = 85;
    if (score >= 90) {
    }
    else if (score >= 80) {
    }
    else if (score >= 70) {
    }
    else if (score >= 60) {
    }
    else {
    }
    std::cout << to_str("Score ") << to_str(score) << to_str(" is grade ") << to_str(grade) << std::endl;
    std::cout << "\n=== Multiple inline conditions ===" << std::endl;
    auto x = 15;
    auto result = (;
    );
    std::cout << to_str("x=") << to_str(x) << to_str(" is ") << to_str(result) << std::endl;
    std::cout << "\n=== Loop with continue ===" << std::endl;
    for (int i = 0; i < 10; ++i) {
    }
    std::cout << "\n=== Enumerate with condition ===" << std::endl;
    auto words = std::vector<std::string>{"apple", "banana", "cherry", "date"};
    for (size_t i = 0; i < words.size(); ++i) {
        auto word = words[i];
    }
    std::cout << "\n=== try-except-else-finally ===" << std::endl;
    divide(10, 2);
    std::cout << std::endl;
    divide(10, 0);
    std::cout << "\n=== Nested try-except ===" << std::endl;
    {
    }
    if (__fluxus_exc) { auto e = 0;
    }
    std::cout << "\n=== Multiple exceptions ===" << std::endl;
    process_value(0);
    process_value(5);
    std::cout << "\n=== All control flow tests completed ===" << std::endl;
    return 0;
}

